main.go

```diff
@@ -183,43 +203,97 @@ func main() {
 	}
 	// +kubebuilder:scaffold:builder
 
+	leaderElectionErr := make(chan error)
+	leaderElectionContext, cancelLeaderElection := context.WithCancel(context.Background())
+	go func() {
+		setupLog.Info("starting leader election manager")
+		leaderElectionErr <- leaderElectionMgr.Start(leaderElectionContext)
+	}()
+
+	select {
+	case <-leaderElectionMgr.Elected():
+	case err := <-leaderElectionErr:
+		setupLog.Error(err, "Leader Election Manager error")
+		os.Exit(1)
+	}
+
 	// Create a default SriovNetworkNodePolicy
 	err = createDefaultPolicy(kubeClient)
 	if err != nil {
 		setupLog.Error(err, "unable to create default SriovNetworkNodePolicy")
+		cancelLeaderElection()
+		<-leaderElectionErr
 		os.Exit(1)
 	}
 
 	// Create default SriovOperatorConfig
 	err = createDefaultOperatorConfig(kubeClient)
 	if err != nil {
 		setupLog.Error(err, "unable to create default SriovOperatorConfig")
+		cancelLeaderElection()
+		<-leaderElectionErr
 		os.Exit(1)
 	}
 
-	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
-		setupLog.Error(err, "unable to set up health check")
-		os.Exit(1)
-	}
-	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
-		setupLog.Error(err, "unable to set up ready check")
-		os.Exit(1)
-	}
+	stopSignalCh := ctrl.SetupSignalHandler()
 
-	stopCh := ctrl.SetupSignalHandler()
+	globalManagerErr := make(chan error)
+	globalManagerCtx, globalManagerCancel := context.WithCancel(context.Background())
 	go func() {
-		if err := mgrGlobal.Start(stopCh); err != nil {
-			setupLog.Error(err, "Manager Global exited non-zero")
-			os.Exit(1)
-		}
+		setupLog.Info("starting global manager")
+		globalManagerErr <- mgrGlobal.Start(globalManagerCtx)
 	}()
 
-	// Remove all finalizers after controller is shut down
-	defer utils.Shutdown()
+	namespacedManagerErr := make(chan error)
+	namespacedManagerCtx, namespacedManagerCancel := context.WithCancel(context.Background())
+	go func() {
+		setupLog.Info("starting namespaced manager")
+		namespacedManagerErr <- mgr.Start(namespacedManagerCtx)
+	}()
+
+	select {
+	// Wait for a stop signal
+	case <-stopSignalCh.Done():
+		setupLog.Info("Stop signal received")
+
+		globalManagerCancel()
+		namespacedManagerCancel()
+		<-globalManagerErr
+		<-namespacedManagerErr
+
+		utils.Shutdown()
+		cancelLeaderElection()
+		<-leaderElectionErr
+
+	case err := <-leaderElectionErr:
+		setupLog.Error(err, "Leader Election Manager error")
+		globalManagerCancel()
+		namespacedManagerCancel()
+		<-globalManagerErr
+		<-namespacedManagerErr
+
+		os.Exit(1)
+
+	case err := <-globalManagerErr:
```

Comment:
I think a shutdown should be included here as well since the namespace manager might be up, and perhaps also in the case above. Not sure since the operator won't be elected anymore for that case.

Reply:
I'm fine with including `utils.Shutdown()` in case of globalManager and namespacedManager error. I wouldn't do it in case of leaderelection error